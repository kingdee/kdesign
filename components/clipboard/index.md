---
category: Components
type: 基础
title: Clipboard
subtitle: 剪贴板
---

剪贴板是一种用于内容复制或剪贴的组件。

## 使用场景

需要复制或剪贴文本或元素内容时。

## API

| 属性      | 说明             | 类型                        | 默认值   | 可选值                   | 版本   |
| --------- | ---------------- | --------------------------- | -------- | ------------------------ | ------ |
| action    | 剪贴板动作       | string                      | `copy`   | `copy` `cut`             | 1.7.37 |
| content   | 剪贴内容         | string                      | `-`      | `-`                      | 1.7.37 |
| className | 样式名           | string                      | `-`      | `-`                      | 1.7.37 |
| disabled  | 剪贴板禁用状态   | boolean                     | `false`  | `true` `false`           | 1.7.37 |
| icon      | 自定义图标       | ReactNode \| false          | -        | -                        | 1.7.37 |
| size      | 剪贴板尺寸       | string                      | `middle` | `small` `middle` `large` | 1.7.37 |
| style     | 剪贴板样式       | CSSProperties               | -        | -                        | 1.7.37 |
| target    | 剪贴目标         | () => HTMLElement \| string | `-`      | `-`                      | 1.7.37 |
| onError   | 剪贴失败时的回调 | (error: Error) => void      | `-`      | `-`                      | 1.7.37 |
| onSuccess | 剪贴成功时的回调 | (content: string) => void   | `-`      | `-`                      | 1.7.37 |
