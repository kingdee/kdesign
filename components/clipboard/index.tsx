import InternalClipboard from './clipboard'

const Clipboard = InternalClipboard

export { ClipboardAction, IClipboardProps } from './clipboard'
export default Clipboard
