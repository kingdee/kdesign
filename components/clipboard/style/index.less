@import '../../style/themes/index';
@import './mixin.less';
@clipboard-prefix-cls: ~'@{kd-prefix}-clipboard';

.@{clipboard-prefix-cls} {
  .clipboard;

  &-size-small {
    .clipboard-size(@clipboard-small-height, @clipboard-small-font-size, @clipboard-small-padding-vertical, @clipboard-small-padding-horizontal);
    &.@{clipboard-prefix-cls}-icon-only {
      .clipboard-icon-only(@clipboard-icon-small-font-size, @clipboard-icon-padding-horizontal);
    }
  }

  &-size-middle {
    .clipboard-size(@clipboard-middle-height, @clipboard-middle-font-size, @clipboard-middle-padding-vertical, @clipboard-middle-padding-horizontal);
    &.@{clipboard-prefix-cls}-icon-only {
      .clipboard-icon-only(@clipboard-icon-middle-font-size, @clipboard-icon-padding-horizontal);
    }
  }

  &-size-large {
    .clipboard-size(@clipboard-large-height, @clipboard-large-font-size, @clipboard-large-padding-vertical, @clipboard-large-padding-horizontal);
    &.@{clipboard-prefix-cls}-icon-only {
      .clipboard-icon-only(@clipboard-icon-large-font-size, @clipboard-icon-padding-horizontal);
    }
  }

  &:not(.@{clipboard-prefix-cls}-icon-only) > i:not(:last-child) {
    margin-right: 2px;
  }
}
