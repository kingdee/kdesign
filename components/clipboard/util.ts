/**
 * 由将被复制的文本为 value 创建一个 HTMLTextAreaElement 容器
 * @param {String} value
 * @return {HTMLTextAreaElement}
 */
const createContainer = (value: string): HTMLTextAreaElement => {
  const isRTL = document.documentElement.getAttribute('dir') === 'rtl'
  const container = document.createElement('textarea')
  // 防止在 ios 上缩放
  container.style.fontSize = '12pt'
  // 重置样式
  container.style.border = '0'
  container.style.padding = '0'
  container.style.margin = '0'
  // 把元素定位在视窗外
  container.style.position = 'absolute'
  container.style[isRTL ? 'right' : 'left'] = '-9999px'
  const yPosition = window.pageYOffset || document.documentElement.scrollTop
  container.style.top = `${yPosition}px`
  // 设置为只读
  container.setAttribute('readonly', '')
  // 将文本赋值给元素
  container.value = value
  return container
}

/**
 * 选中 Element，并返回其选中的文本内容
 * @param {HTMLElement} element
 * @return {string}
 */
const selectElement = (element: HTMLElement): string => {
  let selectedText

  if (element.nodeName === 'SELECT') {
    element.focus()

    selectedText = (element as HTMLSelectElement).value
  } else if (element.nodeName === 'INPUT' || element.nodeName === 'TEXTAREA') {
    const isReadOnly = element.hasAttribute('readonly')

    if (!isReadOnly) {
      element.setAttribute('readonly', '')
    }

    if (element.nodeName === 'INPUT') {
      ;(element as HTMLInputElement).select()
      ;(element as HTMLInputElement).setSelectionRange(0, (element as HTMLInputElement).value.length)
      selectedText = (element as HTMLInputElement).value
    } else {
      ;(element as HTMLTextAreaElement).select()
      ;(element as HTMLTextAreaElement).setSelectionRange(0, (element as HTMLTextAreaElement).value.length)
      selectedText = (element as HTMLTextAreaElement).value
    }

    if (!isReadOnly) {
      element.removeAttribute('readonly')
    }
  } else {
    if (element.hasAttribute('contenteditable')) {
      element.focus()
    }

    // 获取 select
    const selection = window.getSelection()!
    // 创建 range
    const range = document.createRange()

    // 设置 range 的元素
    range.selectNodeContents(element)
    // 重置 select
    selection.removeAllRanges()
    // 设置 select 的 range
    selection.addRange(range)

    // 获取选中的文本
    selectedText = selection.toString()
  }

  return selectedText
}

/**
 * 通过值来复制
 * @param {string} value
 * @return {string}
 */
const copyByValue = (value: string): string => {
  // 以 value 作为文本内容创建一个 HTMLTextAreaElement 临时容器
  const container: HTMLTextAreaElement = createContainer(value)
  // 将临时容器拼接到 body
  document.body.appendChild(container)
  // 选中临时容器，并返回其文本
  const selectedText = selectElement(container)
  try {
    document.execCommand('copy')
  } catch (error) {
    // 不支持 document.execCommand 时的替补方式
    navigator.clipboard.writeText(value).then(null, () => {
      throw new Error(`failed to copy to clipboard of value '${value}'`)
    })
  } finally {
    // 取消焦点
    container.blur()
    // 移除临时容器
    document.body.removeChild(container)
  }
  return selectedText
}

/**
 * 通过元素来复制
 * @param {HTMLElement} element
 * @return {string}
 */
const copyByElement = (element: HTMLElement): string => {
  // 克隆原 element 来进行操作，避免 document.body.contains(element) 返回 false 的情况
  // 在 target.md 中，target 参数用 document.getElementById('div-to-copy') 传参时就会出现返回 false 的情况
  const clonedElement = element.cloneNode(true) as HTMLElement
  // 将克隆出来的 element 拼接到 body 上
  document.body.appendChild(clonedElement)
  // 选中克隆的 element，并返回其文本
  const selectedText = selectElement(clonedElement)
  try {
    document.execCommand('copy')
  } catch (error) {
    // 不支持 document.execCommand 时的替补方式
    navigator.clipboard.writeText(selectedText).then(null, () => {
      throw new Error(`failed to copy to clipboard of element '${element}'`)
    })
  } finally {
    // 取消焦点
    clonedElement.blur()
    // 移除 document.body.contains 返回 false 时临时创建的 element
    document.body.removeChild(clonedElement)
  }
  return selectedText
}

/**
 * 剪贴板复制操作
 * @param {string | HTMLElement} target
 * @return {string}
 */
const onClipboardCopy = (target: string | HTMLElement): string => {
  let selectedText = ''
  if (typeof target === 'string') {
    // 如果传的是字符串，直接通过值来复制
    selectedText = copyByValue(target)
  } else if (
    target instanceof HTMLInputElement &&
    !['text', 'search', 'url', 'tel', 'password'].includes(target.type)
  ) {
    // 如果传的是 input 标签的某些类型，通过值来复制
    selectedText = copyByValue(target.value)
  } else {
    // 其他情况均通过元素来复制
    selectedText = copyByElement(target)
  }
  return selectedText
}

/**
 * 剪贴板剪切操作
 * @param {HTMLElement} target
 * @return {string}
 */
const onClipboardCut = (target: HTMLElement): string => {
  // 选中元素，并返回其文本
  const selectedText = selectElement(target)
  try {
    document.execCommand('cut')
  } catch (error) {
    // 不支持 document.execCommand 时的替补方式
    navigator.clipboard.writeText(selectedText).then(null, () => {
      throw new Error(`failed to copy to clipboard of value '${selectedText}'`)
    })
  } finally {
    // 取消焦点
    target.blur()
  }
  return selectedText
}

export { onClipboardCopy, onClipboardCut }
