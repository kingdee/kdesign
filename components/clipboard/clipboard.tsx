import classNames from 'classnames'
import React, { FunctionComponentElement, useCallback, useContext } from 'react'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'
import { tuple } from '../_utils/type'
import ConfigContext from '../config-provider/ConfigContext'
import Icon from '../icon'
import { onClipboardCopy, onClipboardCut } from './util'

export const ClipboardActions = tuple('copy', 'cut')
export type ClipboardAction = typeof ClipboardActions[number]

export const ClipboardSizes = tuple('large', 'middle', 'small')
export type ClipboardSize = typeof ClipboardSizes[number]

export interface IClipboardProps {
  action?: ClipboardAction // 动作
  content?: string // 剪贴内容
  children?: React.ReactNode // 子元素
  className?: string // 样式名
  disabled?: boolean // 禁用状态
  icon?: React.ReactNode | false // 自定义图标
  prefixCls?: string // 样式前缀
  size?: ClipboardSize // 尺寸
  style?: Record<string, unknown> // 内联样式
  target?: () => HTMLElement | string // 剪贴目标，DOM、id or class
  onError?: (error: Error) => void // 失败的回调
  onSuccess?: (content: string) => void // 成功的回调
}

const InternalClipboard = (props: IClipboardProps, ref: unknown): FunctionComponentElement<IClipboardProps> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  // 剪贴板参数，合并了系统默认参数、用户定义组件统一参数及组件传参
  const clipboardProps = getCompProps('Clipboard', userDefaultProps, props)
  const {
    className,
    prefixCls: customPrefixcls,
    action,
    content,
    children,
    disabled,
    icon,
    size,
    target,
    onError,
    onSuccess,
    ...others
  } = clipboardProps

  devWarning(ClipboardActions.indexOf(action) === -1, 'clipboard', `cannot found clipboard action '${action}'`)
  devWarning(ClipboardSizes.indexOf(size) === -1, 'clipboard', `cannot found clipboard size '${size}'`)

  // 剪贴板样式前缀
  const clipboardPrefixCls = getPrefixCls!(prefixCls, 'clipboard', customPrefixcls)
  const clipboardClass = classNames(clipboardPrefixCls, className, {
    [`${clipboardPrefixCls}-size-${size}`]: size, // 尺寸样式
    [`${clipboardPrefixCls}-icon-only`]: !children && children !== 0 && icon !== null, // 只有图标
  })

  // ref
  const clipboardRef = (ref as any) || React.createRef<HTMLElement>()

  // 如果没有传 icon，则使用默认图标
  // 如果 icon 传 null，则不使用图标
  const iconNode = icon || (icon === undefined ? <Icon type="copy-code" /> : null)

  // 如果传的 children 为 string，则将 children 包装成块标签
  const childrenNode = typeof children === 'string' ? <span>{children}</span> : children

  /**
   * 剪贴板 click 事件处理
   */
  const handleClick = useCallback(() => {
    if (disabled) {
      // 禁用状态不触发点击事件
      return
    }
    let selectedText
    try {
      if (content) {
        // 如果传了 content，则直接复制 content
        selectedText = onClipboardCopy(content)
      } else {
        const targetElement: HTMLElement = typeof target === 'string' ? document.querySelector(target) : target()
        if (!targetElement) {
          // 没有获取到元素，抛出异常
          throw new Error(`target element '${target}' does not exist`)
        } else {
          if (action === 'copy' && targetElement.hasAttribute('disabled')) {
            // 复制被置为 disabled 的元素，抛出异常
            throw new Error(
              `invalid "target" attribute of element '${targetElement}'. Please use "readonly" instead of "disabled" attribute`,
            )
          }
          if (action === 'cut' && (targetElement.hasAttribute('readonly') || targetElement.hasAttribute('disabled'))) {
            // 剪切被置为 readonly 或 disabled 的元素，抛出异常
            throw new Error(
              `invalid "target" attribute of element '${targetElement}'. You can't cut text from elements with "readonly" or "disabled" attributes`,
            )
          }
        }
        // 对 targetElement 的值进行复制或剪切
        selectedText = action === 'copy' ? onClipboardCopy(targetElement) : onClipboardCut(targetElement)
      }
    } catch (error) {
      // 触发失败的回调
      onError?.(error)
      // 退出处理器
      return
    }
    // 触发成功的回调
    onSuccess?.(selectedText)
  }, [action, content, disabled, onError, onSuccess, target])

  return (
    <div className={clipboardClass} disabled={disabled} onClick={handleClick} ref={clipboardRef} {...others}>
      {iconNode}
      {childrenNode}
    </div>
  )
}

const Clipboard = React.forwardRef<unknown, IClipboardProps>(InternalClipboard)
Clipboard.displayName = 'Clipboard'
export default Clipboard
