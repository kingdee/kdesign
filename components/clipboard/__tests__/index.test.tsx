import { mount, render } from 'enzyme'
import React from 'react'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
import Icon from '../../icon'
import { ClipboardAction, ClipboardActions, ClipboardSize, ClipboardSizes } from '../clipboard'
import Clipboard from '../index'

describe('Clipboard', () => {
  // 1. mount test
  describe('1. mount test', () => {
    mountTest(Clipboard)
    ClipboardActions.forEach((action) => {
      mountTest(() => <Clipboard action={action} />)
    })
    ClipboardSizes.forEach((size) => {
      mountTest(() => <Clipboard size={size} />)
    })
  })

  // 2. render test
  describe('2. render test', () => {
    it('renders correctly', () => {
      expect(render(<Clipboard />)).toMatchSnapshot()

      expect(render(<Clipboard style={{ width: 100 }} />)).toMatchSnapshot()

      expect(render(<Clipboard content={'00268'} />)).toMatchSnapshot()

      expect(
        render(
          <Clipboard
            onError={(e) => {
              console.log(e)
            }}
          ></Clipboard>,
        ),
      ).toMatchSnapshot()

      ClipboardActions.forEach((action) => {
        expect(render(<Clipboard action={action} />)).toMatchSnapshot()
      })
      ClipboardSizes.forEach((size) => {
        expect(render(<Clipboard size={size} />)).toMatchSnapshot()
      })
    })
  })

  // 3. warns in component
  describe('3. warns in component', () => {
    it('warns if action is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const action = 'who am I' as any as ClipboardAction
      render(<Clipboard action={action} />)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch(
        "Warning: [kdesign]-clipboard: cannot found clipboard action 'who am I'",
      )
    })

    it('warns if size is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const size = 'who am I' as any as ClipboardSize
      render(<Clipboard size={size} />)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-clipboard: cannot found clipboard size 'who am I'")
    })
  })

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined without errors', () => {
      const wrapper = (
        <Clipboard>
          {null}
          {undefined}
        </Clipboard>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      expect(Clipboard.displayName).toBe('Clipboard')
    })
  })

  // 6. class state
  describe('6. class state', () => {
    it('other properties should be passed', () => {
      const handleCallback = jest.fn()
      const wrapper = mount(
        <Clipboard className="my-class" onSuccess={handleCallback} onError={handleCallback} disabled data-test="test">
          hello world
        </Clipboard>,
      )
      expect(wrapper).toHaveClassName('my-class')
      wrapper.simulate('click')
      expect(handleCallback).not.toHaveBeenCalled()
      expect(wrapper.prop('data-test')).toEqual('test')

      const DefaultClipboard = mount(<Clipboard />)
      expect(DefaultClipboard.find('.kd-clipboard')).toHaveClassName('.kd-clipboard-size-middle')
      expect(DefaultClipboard.find('.kd-clipboard')).toHaveClassName('.kd-clipboard-icon-only')

      ClipboardSizes.forEach((size) => {
        const TestClipboard = mount(<Clipboard size={size}>copy</Clipboard>)
        expect(TestClipboard.find(`.kd-clipboard`)).toHaveClassName(`.kd-clipboard-size-${size}`)
        expect(TestClipboard.find(`.kd-clipboard`)).not.toHaveClassName('.kd-clipboard-icon-only')
      })
    })
  })

  // 7.component interaction(event)
  describe('7.component interaction(event)', () => {
    it('should not clickable when Clipboard is disabled', () => {
      const onSuccess = jest.fn()
      const wrapper = mount(
        <Clipboard disabled content="www.kingdee.com" onSuccess={onSuccess}>
          clipboard
        </Clipboard>,
      )
      wrapper.simulate('click')
      expect(onSuccess).not.toHaveBeenCalledWith()
    })

    it('should trigger error callback ', () => {
      const onError = jest.fn()
      const wrapper = mount(<Clipboard onError={onError}>clipboard</Clipboard>)
      wrapper.simulate('click')
      expect(onError).toHaveBeenCalled()
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const clipboardConfig = {
        compDefaultProps: {
          Clipboard: {
            action: 'cut',
            size: 'large',
            disabled: true,
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={clipboardConfig}>
          <Clipboard>Clipboard Text</Clipboard>
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-clipboard')).toHaveClassName('.kd-clipboard-size-large')
      expect(wrapper.find('.kd-clipboard')).toHaveProp('disabled')
    })
  })

  // 9. ref test
  describe('9. ref test', () => {
    it('should get Clipboard element from ref', () => {
      const ref = React.createRef()
      mount(<Clipboard ref={ref}></Clipboard>)
      expect((ref.current as HTMLElement).classList.contains('kd-clipboard')).toBe(true)
    })
  })

  // 10. api test
  describe('10. api test', () => {
    it('icon api', () => {
      const ref = React.createRef()
      const icon = <Icon type="copy" />
      const wrapper = mount(<Clipboard ref={ref} icon={icon}></Clipboard>)
      expect(wrapper.find('i')).toHaveClassName('kdicon-copy')
    })
  })

  // 11. special case
})
