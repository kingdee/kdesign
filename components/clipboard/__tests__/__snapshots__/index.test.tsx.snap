// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clipboard 2. render test renders correctly 1`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 2`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
  style="width:100px"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 3`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 4`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 5`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 6`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 7`] = `
<div
  class="kd-clipboard kd-clipboard-size-large kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 8`] = `
<div
  class="kd-clipboard kd-clipboard-size-middle kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 2. render test renders correctly 9`] = `
<div
  class="kd-clipboard kd-clipboard-size-small kd-clipboard-icon-only"
>
  <i
    class="kdicon kdicon-copy-code"
  />
</div>
`;

exports[`Clipboard 4. render null or undefined without errors render null or undefined without errors 1`] = `<Clipboard />`;
